packages/firestore/src/remote/datastore.ts

```diff
@@ -158,3 +89,100 @@ export class Datastore {
       });
   }
 }
+
+export function newDatastore(
+  connection: Connection,
+  credentials: CredentialsProvider,
+  serializer: JsonProtoSerializer
+): Datastore {
+  return new DatastoreImpl(connection, credentials, serializer);
+}
+
+export async function invokeCommitRpc(
+  datastore: Datastore,
+  mutations: Mutation[]
+): Promise<MutationResult[]> {
+  debugAssert(
```

Comment:
Is IntelliJ happier if you made this something like this here:

```ts
const datastoreImpl = cast(datastore);
```

and also added a function like this:

```ts
function cast(datastore: Datastore): DatastoreImpl {
  debugAssert(...);
  return datastore as DatastoreImpl;
}
```

Reply:
I just used an explicit cast to a new type, which makes IntelliJ happy.

Reply:
A `cast` function also has the benefit of reducing the boilerplate at the top of every function from 5 lines to 1. The specific text of which function needed DatastoreImpl isn't important to preserve--should we run afoul of this assertion the stack will show with function was invoked improperly.

If you want to avoid the extra function in the emitted code (if that's a problem) we could add a rewrite rule to inline the function call (to just a cast). If `cast` is too ambiguous for the strategy, maybe name it `castDatastoreImpl`,  `recoverDatastoreImpl`, 'toDatastoreImpl`, or `asDatastoreImpl`? This way you could write dumb, unscoped global rules for this (again--if that's even worthwhile).

Reply:
Not 100% convinced that this effort was worth it (since all these extra lines are stripped in production), but I added a `debugCast` function that not only taught me a lot about TypeScript and constructor signatures but also can be used everywhere else in the code.

Reply:
Agreed that this makes little to no difference in the production code (after stripping). Where it makes a difference is in the readability of this required transformation. Having the same 5-line incantation at the top of nearly every function would have made my eyes bleed :-).